!!! BUGFIX: Always unset node property, if set to null

mhsdesign opened the pull request:
Previously we made a distinction between unsetting a node property and setting it to null on the write side.

There was setting to null `SerializedPropertyValues::fromArray(['tagName' => new SerializedPropertyValue(null, 'string')])` and unset `SerializedPropertyValues::fromArray(['tagName' => null])`

Unsetting a property would only be possible via php api by creating a custom `SerializedPropertyValues`.

The read site treats it the same and returns true in hasProperty only if there is a non null value in the database.

As discussed with @bwaidelich 

>  I would recommend to **not make a difference whether a property wasn't set initially or whether it was removed lateron**.
While it might be useful to expose this information at some point, we should do so _explicitly_ and not via the `null`-behavior of some adapter.

Previously if a property tagName was set to `null`

the `NodePropertiesWereSet` event would contain
```
"propertyValues":{"tagName":{"value":null,"type":"string"}}
```
and the properties of the node in the projection as well
```
"tagName":{"value":null,"type":"string"}
```

-----

which this change the the `NodePropertiesWereSet` event would contain:
```
"propertyValues":{"tagName":null}
```
and the properties of the node in the projection would show no hint of this property.

-------

Our logic to decide between unsetting properties and setting them to null was spread through a lot of classes and is a sheer wonder that it worked :D Now with the instruction of the `UnsetPropertyValue`, it is much more obvious how the behaviour works. 

solves: #4467

> [!IMPORTANT]  
> This change is breaking to your existing content repository data. To migrate your events please execute:
> ```sh
> flow migrateevents:migratePropertiesToUnset
> flow cr:projectionReplay contentGraph
> ```
> otherwise you will face this error `SerializedPropertyValue::__construct(): Argument #1 ($value) must be of type ArrayObject|array|string|int|float|bool, null given`

mhsdesign committed:
TASK: Always unset node property, if set to null

bwaidelich commented:
From the description and the code itself, I don't really understand this change. But I have the strong feeling that is should be backed by a test

mhsdesign commented:
Yes i see, sorry for being so spare about the details.

Id like to "quickly" chat with you about this to structure my mind - after that i could probably write it down more easily, and we could also discuss if my expectations are correct.

bwaidelich commented:
> Id like to "quickly" chat with you about this to structure my mind

Those quotes around "quickly" make me suspicious :)
No, sure thing, let's talk about this sometimes next week!

bwaidelich commented:
I added some thoughts to https://github.com/neos/neos-development-collection/issues/4304#issuecomment-1604042784

skurfuerst reviewed (approved):
looks good :) wonder why the stuff before was done ;)

mhsdesign commented:
Now i only have to come up with a test for this ;) (Thats why its WIP - and also need to rebase it)

mhsdesign commented on Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php:
```diff
@@ -44,8 +44,8 @@ public function serializePropertyValues(
         foreach ($propertyValuesToWrite->values as $propertyName => $propertyValue) {
-            if (!isset($nodeType->getProperties()[$propertyName]) && $propertyValue === null) {
-                // The property is undefined and the value null => we want to remove it, so we set it to null
+            if ($propertyValue === null) {
```
`!isset($nodeType->getProperties()[$propertyName])` is at that point always `false` because we only allow to setProperties that actually exist in the current schema ([validateProperties](https://github.com/neos/neos-development-collection/blob/9c8f3077b337ae38556d43cbd75142ee1484944a/Neos.ContentRepository.Core/Classes/Feature/NodeCreation/NodeCreation.php#L117-L130)) - thats why the the code path was previously never taken.

mhsdesign committed:
TASK: Obvious unset and null property handling

mhsdesign committed:
TASK: Fix tests because setting to `null` will be unset

Previously setting `null` to a string property would lead to an empty string in the database. But now the property will be unset.

mhsdesign commented:
As we discussed previously it is an implementation detail of the adapter to save null values to the database or not, thats why its a bit hard to come up with a test? And as everything passes that seems proof enough this refactoring is correct?

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValue.php:
```diff
@@ -18,16 +18,17 @@
      */
     public function __construct(
-        public readonly int|float|string|bool|array|\ArrayObject|null $value,
+        public readonly int|float|string|bool|array|\ArrayObject $value,
```
its _might_ be uncool to have this signature and always having to check the value before against null.
We could also have a static factory that returns either `SerializedPropertyValue` or `UnsetPropertyValue` if null was accidentally still passed?

mhsdesign replied:
solved via `SerializedPropertyValue::create`

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValues.php:
```diff
@@ -52,10 +49,14 @@ public static function fromArray(array $propertyValues): self
+            if ($propertyValue === null || $propertyValue instanceof UnsetPropertyValue) {
+                $values[$propertyName] = UnsetPropertyValue::get();
             } elseif (is_array($propertyValue)) {
+                if (array_key_exists('value', $propertyValue) && $propertyValue['value'] === null) {
+                    // legacy
+                    $values[$propertyName] = UnsetPropertyValue::get();
+                    continue;
+                }
```
without this upcasting, a replay on older event would lead to:

> SerializedPropertyValue::__construct(): Argument #1 ($value) must be of type ArrayObject|array|string|int|float|bool, null given, called in 

alternatively, we could possibly migrate the events table?

mhsdesign replied:
solved via `SerializedPropertyValue::create`

mhsdesign committed:
TASK: Introduce `SerializedPropertyValue::create`

mhsdesign committed:
TASK: Introduce `SerializedPropertyValues::withoutUnsets`

To make the code more obvious

mhsdesign committed:
TASK: Make sure unset properties never leak into the node

The Node's properties must not reflect that a node's property was unset once.

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValue.php:
```diff
@@ -18,24 +18,40 @@
+     *
+     * @param int|float|string|bool|array<int|string,mixed>|\ArrayObject<int|string,mixed>|null $value
+     */
+    public static function create(
+        int|float|string|bool|array|\ArrayObject|null $value,
+        string $type
+    ): self|UnsetPropertyValue {
+        if ($value === null) {
+            return UnsetPropertyValue::get();
+        }
+        return new self($value, $type);
+    }
```
Do we need to expose the "null"-hack to the new constructor?
I would suggest to change this to:
```suggestion
    public static function create(
        int|float|string|bool|array|\ArrayObject $value,
        string $type
    ): self|UnsetPropertyValue {
        return new self($value, $type);
    }
```

and then...

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/SerializedPropertyValue.php:
```diff
@@ -44,7 +60,7 @@ public static function fromArray(array $valueAndType): self
         }
 
-        return new self($valueAndType['value'], $valueAndType['type']);
+        return self::create($valueAndType['value'], $valueAndType['type']);
```
...adjust this accordingly:
```suggestion
        return $valueAndType['value'] === null ? UnsetPropertyValue::get() : self::create($valueAndType['value'], $valueAndType['type']);
```

bwaidelich reviewed (dismissed):
Sorry for being complicated, but I dislike that one of the highest-level APIs now has this awkward `withoutUnsets()` method.
I am probably missing something but I expected `SerializedProperties` to be just that, a set of serialized properties. But that they would never contain any `null`-values.
I.e. that we would filter them out at _write_ time.

I know, that we need it to _unset_ previously existing properties, but I thought that we could put that logic into `PropertyValuesToWrite` rather than having something like `Node->properties->serialized->withoutUnsets()` which does not make sense to me on the read side at all

mhsdesign commented:
We discussed to introduce a `PropertyNames` dto and a new property in the `SetSerializedProperties` command / event `$propertiesToUnset`.

Additionally we briefly discussed that a property converter for the properties should probably never return `null`, as this would cause confusion about `hasProperty` seeminly not working.

mhsdesign committed:
TASK: Document merging or replace behaviour of Set* commands

mhsdesign committed:
WIP: Introduce $propertiesToUnset to command

mhsdesign committed:
WIP: Introduce $propertiesToUnset to events

mhsdesign committed:
WIP: Introduce `SerializedPropertyValues::unsetProperties`

mhsdesign committed:
WIP: Remove WIP `UnsetPropertyValue`

And make node get property throw if value deserializes to null.

mhsdesign committed:
WIP: Fix `SetSerializedNodeProperties` event

mhsdesign committed:
WIP: Reintroduced initial `EventMigrationService`

During extending phpstan to more packages (https://github.com/neos/neos-development-collection/pull/4650) the migration was dropped.
The code had become already outdated by then as the api was changed a lot.
This service should function as boilerplate for new migrations.

mhsdesign committed:
WIP: Try to fix previous `EventMigrationService`

After https://github.com/neos/neos-development-collection/pull/4405 a lot of things changed.

mhsdesign commented:
So i also just wrote a database migration that will update the events. Adding a php compatibility layer would just permanently introduce complexity and possibly bugs.

The following things have to be migrated:

- `NodePropertiesWereSet` `payload`
- `SetSerializedNodeProperties` in `metadata`
example:
```
"propertyValues":{"tagName":{"value":null,"type":"string"}}
->
"propertyValues":[],"propertiesToUnset":["tagName"]
```

Here we just have to omit the null values because its a new node either way:

- `NodeAggregateWithNodeWasCreated` `payload`
- `CreateNodeAggregateWithNodeAndSerializedProperties` in `metadata`
```
"initialPropertyValues":{"title":{"value":"Blog","type":"string"},"titleOverride":{"value":null,"type":"string"},"uriPathSegment":{"value":"blog","type":"string"}}
->
"initialPropertyValues":{"title":{"value":"Blog","type":"string"},"uriPathSegment":{"value":"blog","type":"string"}}
```
- `CopyNodesRecursively` in `metadata`

```
// somewhere deep recursive in $nodeTreeToInsert
"propertyValues":{"senderName":{"value":null,"type":"string"}}}
-> 
"propertyValues":{}
```



Also the basically impossible `"initialPropertyValues":{"tagName":null}` and `"propertyValues":{"tagName":null` will be migrated.

mhsdesign committed:
TASK: Add migration `flow migrateevents:migratePropertiesToUnset`

mhsdesign committed:
TASK: Fix linter

mhsdesign commented:
This is now ready to be reviewed :D And with the migration mergeable.

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyNames.php:
```diff
@@ -0,0 +1,66 @@
+use Neos\ContentRepository\Core\NodeType\NodeType;
+use Neos\ContentRepository\Core\SharedModel\Node\PropertyName;
+
+/**
+ * Todo move to shared and use {@see PropertyName} ???
+ * @api used as part of commands/events
+ */
+final readonly class PropertyNames implements \JsonSerializable
```
Todo move to shared and use {@see PropertyName} ???

mhsdesign committed:
TASK: Limit cache job to fail fast

in the past the job ran sometimes for 6 hours, which is a waste of resources.

ahaeslich commented:
Just tested the migration on a project with ~100k events. 38 events were migrated, for example:

Before: `"propertyValues":{"image":{"value":null,"type":"Neos\\Media\\Domain\\Model\\ImageInterface"}}`
After: `"propertyValues":[],"propertiesToUnset":["image"]`

ahaeslich commented on Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:
```diff
@@ -0,0 +1,259 @@
+        $streamName = VirtualStreamName::all();
+        $eventStream = $this->eventStore->load($streamName);
+        foreach ($eventStream as $eventEnvelope) {
+            $outputRewriteNotice = fn(string $message) => $outputFn(sprintf('%s@%s %s', $eventEnvelope->event->type->value, $eventEnvelope->sequenceNumber->value, $message));
```
Maybe we can reduce the output. If one has many events, warnings or migration messages are hard to find 😅. This also applies to line 110

mhsdesign replied:
yes i wasnt aware of 100k nodes :O Sorry to your terminal

mhsdesign committed:
TASK Improve output of event migration

mhsdesign committed:
TASK: Move `PropertyNames` next to `PropertyName`

mhsdesign committed:
TASK: Use `PropertyName` inside `PropertyNames`

mhsdesign committed:
TASK: Move `PropertyNames::splitByScope`

mhsdesign committed:
TASK: Don't replay automatically after `flow migrateevents:migratePropertiesToUnset`

mhsdesign commented:
Okay with @nezaniel help i also tested the migration on his starship's event stream (of 30K events) where the migration could successfully migrate `12884` events.

Excerpt:

```
2292@NodeAggregateWithNodeWasCreated Payload: Removed 1 $initialPropertyValues
2301@NodeAggregateWithNodeWasCreated Payload: Removed 5 $initialPropertyValues
2350@NodeAggregateWithNodeWasCreated Payload: Removed 1 $initialPropertyValues
...
77513@NodePropertiesWereSet Payload: Migrated 1 $unsetProperties
77513@NodePropertiesWereSet Metadata: Migrated 1 $unsetProperties
77514@NodePropertiesWereSet Payload: Migrated 1 $unsetProperties
77514@NodePropertiesWereSet Metadata: Migrated 1 $unsetProperties
```

I think having the migration now tested with 100K events, 30K and with the demo, we can be sure that it works.

I could also imagine diffing the content graph node projections for full insurance, but i fear that would be to much data :D

kitsunet commented on Neos.ContentRepository.BehavioralTests/Classes/PhpstanRules/InternalMethodsNotAllowedOutsideContentRepositoryRule.php:
```diff
@@ -40,6 +40,8 @@ public function processNode(Node $node, Scope $scope): array
                 || str_starts_with($scope->getNamespace(), 'Neos\ContentRepository\Export')
                 || str_starts_with($scope->getNamespace(), 'Neos\ContentRepository\LegacyNodeMigration')
                 || str_starts_with($scope->getNamespace(), 'Neos\ContentRepository\StructureAdjustment')
+                // we do some next level stuff in the migrations, and thus we do illegal stuff:
```
can we maybe describe this more matter of factly, because this comment doesn't help me at all to understand what's going on here.

kitsunet commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/ContentStreamForking/ForkContentStreamWithoutDimensions.feature:
```diff
@@ -46,6 +46,7 @@ Feature: ForkContentStream Without Dimensions
       | originDimensionSpacePoint    | {}                                                      |
       | affectedDimensionSpacePoints | [{}]                                                    |
       | propertyValues               | {"text": {"value": "original value", "type": "string"}} |
+      | propertiesToUnset            | {}                                                      |
```
This could default to empty so you don't need to add it to every feature?

mhsdesign replied:
jain. `NodePropertiesWereSet` event publishing is discouraged in our tests and thus we dont need to make it more nice.https://github.com/neos/neos-development-collection/issues/4336

kitsunet commented on Neos.ContentRepository.Core/Classes/Infrastructure/Property/PropertyConverter.php:
```diff
@@ -72,26 +67,35 @@ private function serializePropertyValue(
+            );
+        }
+
+        if ($serializedPropertyValue === null) {
```
Mmm, I find this tricky because I could see edge cases were serialization might result in a NULL value without the user really being able to control that?

mhsdesign replied:
Yes i know. I briefly discussed this with bastian freely after lets see :D

The thing is `node->hasProperty` will become more hard to describe and understand if `node->getProperty` can return `null` values if the node _has_ the property. What do you say?

But in the end we can always loosen this ...

kitsunet replied:
I think the tricky part is that getProperty wil need to be able to return null, think policies, lets say you have an entity that not everyone can see, this would be reflected in the property, and the unserialiser shouldn't be able to unserialize it for some viewers.

mhsdesign replied:
I think we should open up the discussion when this case happens. Being more restrictive here in the first place seems valid?
Having an serializer enforce policies is also weird ....

mhsdesign replied:
Actually I’m sure we need to catch this here 100%

if a serializer failed at write/command time that is okay

But the code below in deserializePropertyValue might apply to your concerns where I still find it confusing … is possible in the existing system for example to entity privilege protect an image or any asset? If so we probably do need to talk about that …

kitsunet replied:
> Having an serializer enforce policies is also weird

Not enforce, but like, an entity privilege would result in a referenced object being null if the user reading the node doesn't have permission to see it.

mhsdesign replied:
Yeah okay, reverted the read side error throwing as this is not handleable by the user. The write side is fine imo and desired.

Discussion how this might affect our api and confuse users can be continued here: https://github.com/neos/neos-development-collection/issues/4911

kitsunet commented on Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyValuesToWrite.php:
```diff
@@ -66,4 +66,9 @@ public function merge(self $other): self
     {
         return new self(array_merge($this->values, $other->values));
     }
+
```
To me the name of the DTO suggests that is already contains only stuff that is writeable, couldn't / shouldn't we just handle this directly in hte constructor (and withValue) such that we discard null values?

mhsdesign replied:
the idea was not to break the high level api, that we allow null values at this point and split it later. We _need_ the information as we access it at some point. Might as well add a `getUnsetPropertyNames` to make it more explicit.

kitsunet replied:
Mmm, I guess, it#s just a weird method name vs DTO name and might not be obvious to anyone that you need to "cast" it with this method in some places... But if we need it

mhsdesign replied:
Only we have to take care of casting it. And that is the tradeoff we discussed in favour of not breaking our api. Otherwise we would need to add a new property in the `SetNodeProperties` command itself.

We _could_ of course add this odd logic and splitting right into the `SetNodeProperties` ... by doing it on `create` or break things harder?

kitsunet replied:
Nah it's fine, I see what you mean

kitsunet replied:
@bwaidelich see this, as you commented the same as me (I hadn't seen your comment). I would also love to have it differnetly but if it's really that complicated?

mhsdesign replied:
Bastis comment was on a previous state, we both worked out this one ...

kitsunet commented on Neos.ContentRepository.Core/Classes/Feature/NodeModification/Dto/PropertyScope.php:
```diff
@@ -58,6 +60,15 @@ public function resolveAffectedOrigins(
         };
     }
 
+    public static function tryFromDeclaration(NodeType $nodeType, PropertyName $propertyName): self
```
How is this related to the PR?

mhsdesign replied:
extracted method to deduplicate code in `splitByScope` and `splitPropertiesToUnsetByScope`

... but ha i see you already noticed :D

kitsunet commented on Neos.ContentRepository.Core/Classes/Feature/NodeModification/NodeModification.php:
```diff
@@ -100,9 +109,30 @@ private function handleSetSerializedNodeProperties(
+            }
+        }
+
+        $propertiesToUnsetByScope = $this->splitPropertiesToUnsetByScope($command->propertiesToUnset, $nodeType);
```
I guess that answers my question above regarding the scope... :)

mhsdesign committed:
TASK: Extract `PropertyValuesToWrite::getPropertiesToUnset`

mhsdesign committed:
TASK: Allow `node->getProperty` to return null by deserializer

As discussed here https://github.com/neos/neos-development-collection/pull/4322#discussion_r1501880426
this _can_ likely happen when entity privileges are used.

mhsdesign committed:
TASK: Document null value handling in `PropertyValuesToWrite`

kitsunet reviewed (approved):
Cool, fine by me now, shall I hit merge?

kitsunet commented:
(I guess we agreed in the weekyl this is fine)

mhsdesign commented:
Yes i even did some additional testing that the command payload is also correctly migrated and a workspace with migrated changes can be merged.

Test cases:

- unset node properties via set serialised properties
- copy nodes which have explicitly unset properties which will also be added to the command
- create node with initial unset property

Each of the scenarios worked when being published after switching the branch and applying the `migrateevents:migratePropertiesToUnset` migration.

mhsdesign commented:
Neos.Demo followup https://github.com/neos/Neos.Demo/pull/191
